src/Symfony/Component/String/BinaryString.php

```diff
@@ -0,0 +1,446 @@
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (null === $string = $replace($fromPattern, $to, $this->string)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Matching failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Matching failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        $str = clone $this;
+        $str->string = $string;
+
+        return $str;
+    }
+
+    public function slice(int $start = 0, int $length = null): parent
+    {
+        $str = clone $this;
+        $str->string = (string) substr($this->string, $start, $length ?? \PHP_INT_MAX);
+
+        return $str;
+    }
+
+    public function splice(string $replacement, int $start = 0, int $length = null): parent
+    {
+        $str = clone $this;
+        $str->string = substr_replace($this->string, $replacement, $start, $length ?? \PHP_INT_MAX);
+
+        return $str;
+    }
+
+    public function split(string $delimiter, int $limit = null, int $flags = null): array
+    {
+        if (1 > $limit = $limit ?? \PHP_INT_MAX) {
+            throw new InvalidArgumentException('Split limit must be a positive integer.');
+        }
+
+        if ('' === $delimiter) {
+            throw new InvalidArgumentException('Split delimiter is empty.');
+        }
+
+        if (null !== $flags) {
+            return parent::split($delimiter, $limit, $flags);
+        }
+
+        $str = clone $this;
+        $chunks = $this->ignoreCase
+            ? preg_split('{'.preg_quote($delimiter).'}iD', $this->string, $limit)
+            : explode($delimiter, $this->string, $limit);
+
+        foreach ($chunks as &$chunk) {
+            $str->string = $chunk;
+            $chunk = clone $str;
+        }
+
+        return $chunks;
+    }
+
+    public function startsWith($prefix): bool
+    {
+        if ($prefix instanceof parent) {
+            $prefix = $prefix->string;
+        } elseif (!\is_string($prefix)) {
+            return parent::startsWith($prefix);
+        }
+
+        return '' !== $prefix && 0 === ($this->ignoreCase ? stripos($this->string, $prefix) : strpos($this->string, $prefix));
+    }
+
+    public function title(bool $allWords = false): parent
+    {
+        $str = clone $this;
+        $str->string = $allWords ? ucwords($str->string) : ucfirst($str->string);
+
+        return $str;
+    }
+
+    public function toUtf8(string $fromEncoding = null): Utf8String
+    {
+        $u = new Utf8String();
+
+        if (null === $fromEncoding && preg_match('//u', $this->string)) {
```

Comment:
Should we not also check here if `$fromEncoding` is `utf-8`?

Reply:
That's on purpose: when one is explicit about the input encoding, the code has to obey.

Reply:
I mean if the expression `preg_match('//u', $this->string)` is `true` does that not mean the string is already in utf-8 and we can exit early?

I think the above if could be something like
```
if ((null === $fromEncoding || 'utf-8' === $fromEncoding) && preg_match('//u', $this->string)) {
```

Reply:
I get this is what you meant. The code should not magically decide not to obey because it is too clever. If one calls for converting a string from any-encoding while the string is already in UTF-8 *apparently*, the code must still obey.

Reply:
That will reult in calling `mb_convert_encoding($this->string, 'UTF-8', 'UTF-8');`. Is that useful?
Will not that be just a waste of computer cycles?

Reply:
The "user" or "programmer" has in that case explicitly said the `$fromEncoding` is `utf-8`.
I think that is all the more reason to do nothing, than if `$fromEncoding` is empty.

What can be questioned here, on the grounds of "the code being too clever", is that when `$fromEncoding` is empty it skips calling `mb_convert_encoding` if it detects that it probably already is `utf-8`.

Reply:
OK, got it, sorry for the slowness :)
I also made the conversation stricter, so that an exception is now thrown if the source is not valid for the provided `$fromEncoding`.